spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpRequest.java

```diff
@@ -0,0 +1,174 @@
+
+import java.net.HttpCookie;
+import java.net.URI;
+import java.util.Collection;
+import java.util.function.Function;
+
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.reactive.client.ContentChunk;
+import org.eclipse.jetty.reactive.client.ReactiveRequest;
+import org.eclipse.jetty.util.Callback;
+import org.reactivestreams.Publisher;
+import reactor.core.Exceptions;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import org.springframework.core.io.buffer.DataBuffer;
+import org.springframework.core.io.buffer.DataBufferFactory;
+import org.springframework.core.io.buffer.DataBufferUtils;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.MediaType;
+import org.springframework.util.Assert;
+
+/**
+ * {@link ClientHttpRequest} implementation for the Jetty HTTP client.
+ *
+ * @author Sebastien Deleuze
+ * @since 5.1
+ */
+class JettyClientHttpRequest extends AbstractClientHttpRequest {
+
+	private final Request jettyRequest;
+
+	private final DataBufferFactory bufferFactory;
+
+	private ReactiveRequest reactiveRequest;
+
+	private Mono<ClientHttpResponse> response;
+
+
+	public JettyClientHttpRequest(Request jettyRequest, DataBufferFactory bufferFactory) {
+		this.jettyRequest = jettyRequest;
+		this.bufferFactory = bufferFactory;
+	}
+
+
+	@Override
+	protected void applyHeaders() {
+		HttpHeaders headers = getHeaders();
+		headers.entrySet().forEach(e -> e.getValue().forEach(v -> this.jettyRequest.header(e.getKey(), v)));
+		if (!headers.containsKey(HttpHeaders.ACCEPT)) {
+			this.jettyRequest.header(HttpHeaders.ACCEPT, "*/*");
+		}
+	}
+
+	@Override
+	protected void applyCookies() {
+		getCookies().values().stream().flatMap(Collection::stream)
+				.map(cookie -> new HttpCookie(cookie.getName(), cookie.getValue()))
+				.forEach(this.jettyRequest::cookie);
+	}
+
+	@Override
+	public HttpMethod getMethod() {
+		HttpMethod method = HttpMethod.resolve(this.jettyRequest.getMethod());
+		Assert.state(method != null, "Method must not be null");
+		return method;
+	}
+
+	@Override
+	public URI getURI() {
+		return this.jettyRequest.getURI();
+	}
+
+	@Override
+	public DataBufferFactory bufferFactory() {
+		return this.bufferFactory;
+	}
+
+	@Override
+	public Mono<Void> writeWith(Publisher<? extends DataBuffer> publisher) {
+		String contentType = jettyRequest.getHeaders().contains(HttpHeader.CONTENT_TYPE) ?
+				jettyRequest.getHeaders().getField(HttpHeader.CONTENT_TYPE).getValue():
+				MediaType.APPLICATION_OCTET_STREAM_VALUE;
+		Publisher<ContentChunk> chunks = Flux.from(publisher).map(buffer -> new ContentChunk(buffer.asByteBuffer(), new Callback() {
+			@Override
+			public void succeeded() {
+				DataBufferUtils.release(buffer);
+			}
+
+			@Override
+			public void failed(Throwable x) {
+				DataBufferUtils.release(buffer);
+				throw Exceptions.propagate(x);
+			}
+		}));
```

Comment:
This could be extracted:
```
private ContentChunk toContentChunk(DataBuffer buffer) {
	return new ContentChunk(buffer.asByteBuffer(), new Callback() {
		@Override
		public void succeeded() {
			DataBufferUtils.release(buffer);
		}

		@Override
		public void failed(Throwable x) {
			DataBufferUtils.release(buffer);
			throw Exceptions.propagate(x);
		}
	});
}
```

Then it becomes:
```
Publisher<ContentChunk> chunks = Flux.from(publisher).map(this::toContentChunk);
```
And the writeAndFlushWith method below does the same AFAICS.